### DISCUSSION TITLE
improve filename encoding on win32

### DISCUSSION THREAD

**Discussion by Risiamu** [libvips/libvips#2068] [Ideas]
**Describe the bug**

In `D:\IMAGES\Screenshots` folder i have a set of images called `屏幕截图(15)`. 
![图片](https://user-images.githubusercontent.com/30680793/88469615-136f7580-cf26-11ea-9683-2c4f432a10c3.png)

If i run a batch operation on them like this `ls|%{vips heifsave $_.Name "$($_.Basename).heic"}` i got the following resulted files:
![图片](https://user-images.githubusercontent.com/30680793/88469636-50d40300-cf26-11ea-9f10-cc64dcac5649.png)

I ran a few test in Vscode and find out it seems that utf-8 string `屏幕截图(15)` was directly encoded into GB 2312 `灞忓箷鎴�鍥�(13)`.

This could cause resulted files with Chinese directory redirected into non-existence directory.

For example:

With `ls | %{$_.DirectoryName}` i can get the output showing the directory.

`D:\IMAGES\[第01巻]第01巻第01巻\第01巻第01巻第01巻`.

If I tried to encode it into HEIC using vips.exe using 
`vips heifsave "第01巻第01巻第01巻第01巻第01巻第01巻.png"  "第01巻第01巻第01巻第01巻第01巻第01巻.heic"`

I am running this command in `D:\IMAGES\[第01巻]第01巻第01巻\第01巻第01巻第01巻` 
where the `第01巻第01巻第01巻第01巻第01巻第01巻.png` is stored.

I get no error ouput nor the resulted file.
![图片](https://user-images.githubusercontent.com/30680793/88469580-afe54800-cf25-11ea-8502-d2958b9e1c00.png)

**To Reproduce**
I am not really sure is this reproducible on Machine running with UTF-8 encoding.
I will run some other tests later, see if i can locate the problem.
I uploaded those file with Chinese names here.  
[Screenshots.zip](https://github.com/libvips/libvips/files/4977025/Screenshots.zip)
With any luck, you will be able to reproduce the issue on those files.

**Environment**
 - OS: Win10 (With GB 2312 Encoding, default encoding for Win10 with Chinese choosed as initial language)
 - Vips: 8.9

**Additional context**
I have encountered similiar issue with FFprobe and Mediainfo, but the problem back then is about characters misencoded into UTF-8, not from it. By adding '[Console]::OutputEncoding = [System.Text.Encoding]::UTF8;' in my profile the problem was sloved.

Let me know if i can provide other useful information .

**Comment by Risiamu** [comment]
For Chinese users:
国人用户可以启用这个设置

![图片](https://user-images.githubusercontent.com/30680793/88470218-6f3cfd00-cf2c-11ea-9533-558422cb563b.png)


I tried setting 
```
[Console]::OutputEncoding = [System.Text.Encoding]::UTF8;
[Console]::InputEncoding = [System.Text.Encoding]::UTF8;
```
In my powershell profile, but didn't work.
Lucky, changing the whole system to utf-8 worked.
It seems that vips.exe arbitrarily treat output string with "system default encoding".

Character encoding is propably a complex issues. 
Since this issue is practically addressed for me, feel free to close if there is no further discussion needed.

**Comment by jcupitt** [comment]
Hello @GrandArth,

Yes, libvips started out on Linux, which is just utf-8 everywhere, so I just ignored these issues (mostly). 

libvips uses glib for many file-related operations, and glib is also utf-8 everywhere. The only gotcha is that you have to translate to and from the filename encoding when using native APIs. This has a summary:

https://developer.gnome.org/glib/stable/glib-Character-Set-Conversion.html

So libvips should add calls to `g_filename_to_utf8()` and `g_filename_from_utf8()` whenever it passes filename to and from win32 calls, eg. `vips_popenf()`.

Let's tag this as an enhacement for 8.11.

**Comment by jcupitt** [comment]
Here's a previous attempt to fix this: https://github.com/libvips/libvips/issues/294

There are obviously still a few problems.

**Comment by Risiamu** [comment]
I see, thx for the reply.
